Index: a/arch/arm/mach-str8100/str8100_setup.c
===================================================================
--- a/arch/arm/mach-str8100/str8100_setup.c 2012-01-19 21:25:15.459508039 +1100
+++ b/arch/arm/mach-str8100/str8100_setup.c	2012-01-19 21:42:26.840587235 +1100
@@ -371,24 +371,30 @@
 	{
 		.name =		"ALL",
 		.offset =	CONFIG_ARMBOOT_OFFSET,
-		.size =		0x800000-CONFIG_ARMBOOT_OFFSET,
+		.size =		CONFIG_FLASH_SIZE-CONFIG_ARMBOOT_OFFSET,
 	},{
 		.name =		"ARMBOOT",
 		.offset =	CONFIG_ARMBOOT_OFFSET,
 		.size =		CONFIG_KERNEL_OFFSET-CONFIG_ARMBOOT_OFFSET,
 	},{
-		.name =		"Linux Kernel",
+		.name =		"Kernel",
 		.offset =	CONFIG_KERNEL_OFFSET,
 		.size =		CONFIG_ROOTFS_OFFSET-CONFIG_KERNEL_OFFSET,
 	},{
-		.name =		"MTD Disk1",
+		.name =		"RootFS",
 		.offset =	CONFIG_ROOTFS_OFFSET,
 		.size =		CONFIG_CFG_OFFSET-CONFIG_ROOTFS_OFFSET,
 	},{
-		.name =		"MTD Disk2",
+		.name =		"Configs",
 		.offset =	CONFIG_CFG_OFFSET,
-		.size =		0x800000-CONFIG_CFG_OFFSET,
+		.size =		CONFIG_UBOOT_ENV_OFFSET-CONFIG_CFG_OFFSET,
 	}
+   ,{
+		.name =		"UBootEnv",
+		.offset =	CONFIG_UBOOT_ENV_OFFSET,
+		.size =		CONFIG_FLASH_SIZE-CONFIG_UBOOT_ENV_OFFSET,
+	}
+
 };
 
 static struct flash_platform_data str8100_spi_flash_data = {
@@ -395,7 +395,7 @@
 	.name = "m25p80",
 	.parts = str8100_spi_flash_partitions,
 	.nr_parts = ARRAY_SIZE(str8100_spi_flash_partitions),
-	.type = "m25p64",
+	.type = "m25p128",
 };
 #endif

Index: a/arch/arm/mach-str8100/Kconfig
===================================================================
--- a/arch/arm/mach-str8100/Kconfig	2012-01-20 06:24:31.113133868 +1100
+++ b/arch/arm/mach-str8100/Kconfig	2012-01-20 06:27:12.405593404 +1100
@@ -87,29 +87,44 @@
 	bool "STR8100 flash partition setting"
 
 if STR8100_FLASH_PART
-	config ARMBOOT_OFFSET
-	hex "ARMBOOT OFFSET"
-	default 0x0
-	help
-	  The armboot start offset in flash layout
-
-	config KERNEL_OFFSET
-	hex "KERNEL OFFSET"
-	default 0x40000
-	help
-	  The kernel start offset in flash layout
-
-	config ROOTFS_OFFSET
-	hex "ROOTFS OFFSET"
-	default 0x140000
-	help
-	  The rootfs start offset in flash layout
-
-	config CFG_OFFSET
-	hex "CFG OFFSET"
-	default 0x7f0000
-	help
-	  The cfg start offset in flash layout
+	config FLASH_SIZE 
+	hex "FLASH SIZE" 
+	default 0x1000000 
+	help 
+	  The armboot start offset in flash layout 
+	
+	config ARMBOOT_OFFSET 
+	hex "ARMBOOT OFFSET" 
+	default 0x0 
+	help 
+	  The armboot start offset in flash layout 
+	
+	config KERNEL_OFFSET 
+	hex "KERNEL OFFSET" 
+	default 0x40000 
+	default 0x20000 
+	help 
+	  The kernel start offset in flash layout 
+	
+	config ROOTFS_OFFSET 
+	hex "ROOTFS OFFSET" 
+	default 0x140000 
+	default 0x1400000 
+	help 
+	  The rootfs start offset in flash layout 
+	
+	config CFG_OFFSET 
+	hex "CFG OFFSET" 
+	default 0x7f0000 
+	default 0xEA0000 
+	help 
+	  The cfg start offset in flash layout 
+	
+	config UBOOT_ENV_OFFSET 
+	hex "UBOOT ENV OFFSET" 
+	default 0xF80000 
+	help 
+	  The cfg 
 endif


Index: a/drivers/mtd/devices/m25p80.c
===================================================================
--- a/drivers/mtd/devices/m25p80.c	2012-01-20 07:19:51.191814123 +1100
+++ b/drivers/mtd/devices/m25p80.c	2012-01-20 07:20:19.991658101 +1100
@@ -414,6 +414,7 @@
 	{ "m25p16", 0x14, 0x0000, 64 * 1024, 32 },
 	{ "m25p32", 0x15, 0x0000, 64 * 1024, 64 },
 	{ "m25p64", 0x16, 0x2017, 64 * 1024, 128 },
+	{ "m25p128", 0x16, 0x2018, 256 * 1024, 64 },
 };
 
 /*
