diff -u a/Makefile b/Makefile
--- a/Makefile	2012-01-15 21:58:18.874499098 +1100
+++ b/Makefile	2012-01-16 07:05:22.161294898 +1100
@@ -438,7 +438,7 @@
 include $(srctree)/arch/$(ARCH)/Makefile
 export KBUILD_DEFCONFIG
 
-config %config: scripts_basic outputmakefile FORCE
+%config: scripts_basic outputmakefile FORCE
 	$(Q)mkdir -p include/linux
 	$(Q)$(MAKE) $(build)=scripts/kconfig $@
 	$(Q)$(MAKE) -C $(srctree) KBUILD_SRC= .kernelrelease
